From 8ef0e973b8f1d1a68b527a4409ff754cf0f6d428 Mon Sep 17 00:00:00 2001
From: JW Wang <wangchun@google.com>
Date: Tue, 5 Apr 2022 11:18:35 +0530
Subject: [PATCH 1/3] [PATCH] Don't abandon child sessions (1/n)

It will throw if abandon() is called on a child session.

Bug: 211944991
Bug: 67862680
Test: to be added
Change-Id: Ib0ba9f3786dda2d3174f3ea8c65d1061a3fcb586
Merged-In: Ib0ba9f3786dda2d3174f3ea8c65d1061a3fcb586
(cherry picked from commit 8b67e7db79d15ab448ae8f00b40a4a350ab3d537)
(cherry picked from commit c685f8b19adcec0dc49ffaa1e94d7caa4f9d05ba)
Merged-In:Ib0ba9f3786dda2d3174f3ea8c65d1061a3fcb586
---
 .../com/android/server/pm/PackageInstallerService.java | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/services/core/java/com/android/server/pm/PackageInstallerService.java b/services/core/java/com/android/server/pm/PackageInstallerService.java
index d4b98438d603..5a3779c4f5c4 100644
--- a/services/core/java/com/android/server/pm/PackageInstallerService.java
+++ b/services/core/java/com/android/server/pm/PackageInstallerService.java
@@ -347,9 +347,13 @@ public class PackageInstallerService extends IPackageInstaller.Stub implements
                 final long age = currentTimeMillis - session.createdMillis;
                 if (age >= MAX_SESSION_AGE_ON_LOW_STORAGE_MILLIS) {
                     // Aggressively close old sessions because we are running low on storage
-                    // Their staging dirs will be removed too
-                    session.abandon();
-                } else {
+                    // Their staging dirs will be removed too                    
+                    PackageInstallerSession root = !session.hasParentSessionId()
+                           ? session : mSessions.get(session.getParentSessionId());
+                    if (!root.isDestroyed()) {
+                        root.abandon();
+                    }
+                } else {        
                     // Session is new enough, so it deserves to be kept even on low storage
                     unclaimedStagingDirsOnVolume.remove(session.stageDir);
                 }
-- 
2.17.1

